File: java/lib/src/main/java/io/openweb3/xwebhook/BackgroundTask.java
package io.openweb3.xwebhook;

import io.openweb3.xwebhook.exceptions.ApiException;
import io.openweb3.xwebhook.internal.api.BackgroundTasksApi;
import io.openweb3.xwebhook.models.BackgroundTaskOut;
import io.openweb3.xwebhook.models.ListResponseBackgroundTaskOut;

public final class BackgroundTask {
    private final BackgroundTasksApi api;

    BackgroundTask() {
        api = new BackgroundTasksApi();
    }

    public ListResponseBackgroundTaskOut list(final BackgroundTaskListOptions options) throws ApiException {
        try {
            return api.listBackgroundTasks(options.getStatus(), options.getTask(), options.getLimit(),
                    options.getIterator(),
                    options.getOrder());
        } catch (io.openweb3.xwebhook.internal.ApiException e) {
            throw Utils.wrapInternalApiException(e);
        }
    }

    public BackgroundTaskOut get(final String taskId) throws ApiException {
        try {
            return api.getBackgroundTask(taskId);
        } catch (io.openweb3.xwebhook.internal.ApiException e) {
            throw Utils.wrapInternalApiException(e);
        }
    }
}

File: kotlin/lib/src/main/kotlin/BackgroundTask.kt
package io.openweb3.xwebhook.kotlin

import io.openweb3.xwebhook.kotlin.exceptions.ApiException
import io.openweb3.xwebhook.kotlin.internal.apis.BackgroundTasksApi
import io.openweb3.xwebhook.kotlin.models.BackgroundTaskOut
import io.openweb3.xwebhook.kotlin.models.ListResponseBackgroundTaskOut

class BackgroundTask internal constructor(token: String, options: XwebhookOptions) {
    private val api = BackgroundTasksApi(options.serverUrl)

    init {
        api.accessToken = token
        api.userAgent = options.getUA()
        options.initialRetryDelayMillis?.let { api.initialRetryDelayMillis = it }
        options.numRetries?.let { api.numRetries = it }
    }

    suspend fun list(options: BackgroundTaskListOptions = BackgroundTaskListOptions()): ListResponseBackgroundTaskOut {
        try {
            return api.listBackgroundTasks(options.status, options.task, options.limit, options.iterator, options.order)
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun get(taskId: String): BackgroundTaskOut {
        try {
            return api.getBackgroundTask(taskId)
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }
}
